File: iosApp/LucidDreaming/ViewModels/ModulesViewModel.swift
import Foundation
import Combine
import Shared

class ModulesViewModel: ObservableObject {
    @Published var modulesResponse: ModuleResponse? = nil
    @Published var isLoading: Bool = false
    @Published var error: String? = nil
    
    private let repository = GameInfoRepository()
    private let settingsPreferences = iOSSsettingsPreferences()
    private var cancellables = Set<AnyCancellable>()
    private var autoRefreshEnabled = false
    private var refreshTask: Task<Void, Never>? = nil
    
    func loadModules() {
        isLoading = true
        error = nil
        
        Task {
            do {
                let result = try await repository.getModules()
                switch result {
                case .success(let response):
                    modulesResponse = response
                case .failure(let error):
                    self.error = error.localizedDescription
                }
                
                isLoading = false
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }
    
    func toggleModule(moduleName: String) {
        Task {
            do {
                let result = try await repository.toggleModule(moduleName: moduleName, action: "toggle")
                switch result {
                case .success(let module):
                    // Reload modules to update the UI
                    loadModules()
                case .failure(let error):
                    self.error = error.localizedDescription
                }
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
    
    func enableModule(moduleName: String) {
        Task {
            do {
                let result = try await repository.toggleModule(moduleName: moduleName, action: "enable")
                switch result {
                case .success(let module):
                    loadModules()
                case .failure(let error):
                    self.error = error.localizedDescription
                }
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
    
    func disableModule(moduleName: String) {
        Task {
            do {
                let result = try await repository.toggleModule(moduleName: moduleName, action: "disable")
                switch result {
                case .success(let module):
                    loadModules()
                case .failure(let error):
                    self.error = error.localizedDescription
                }
            } catch {
                self.error = error.localizedDescription
            }
        }
    }
    
    func clearError() {
        error = nil
    }
    
    func startAutoRefresh() {
        guard !autoRefreshEnabled else { return }
        autoRefreshEnabled = true
        
        refreshTask = Task {
            while autoRefreshEnabled {
                let settings = await settingsPreferences.settings.first(where: { _ in true })
                let interval = settings?.refreshInterval ?? 10
                
                try? await Task.sleep(nanoseconds: UInt64(interval) * 1_000_000_000)
                
                if autoRefreshEnabled {
                    loadModules()
                }
            }
        }
    }
    
    func stopAutoRefresh() {
        autoRefreshEnabled = false
        refreshTask?.cancel()
        refreshTask = nil
    }
}

File: androidApp/src/main/java/com/luciddreaming/android/viewmodel/ModulesViewModel.kt
package com.luciddreaming.android.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.luciddreaming.shared.data.model.Module
import com.luciddreaming.shared.data.model.ModuleResponse
import com.luciddreaming.shared.data.preferences.AndroidSettingsPreferences
import com.luciddreaming.shared.data.repository.GameInfoRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.delay

class ModulesViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = GameInfoRepository()
    private val settingsPreferences = AndroidSettingsPreferences(application)

    private val _modulesResponse = MutableStateFlow<ModuleResponse?>(null)
    val modulesResponse: StateFlow<ModuleResponse?> = _modulesResponse.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private var autoRefreshEnabled = false

    fun loadModules() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null

            val result = repository.getModules()
            if (result.isSuccess) {
                _modulesResponse.value = result.getOrNull()
            } else {
                _error.value = result.exceptionOrNull()?.message ?: "Failed to load modules"
            }

            _isLoading.value = false
        }
    }

    fun toggleModule(moduleName: String) {
        viewModelScope.launch {
            val result = repository.toggleModule(moduleName, "toggle")
            if (result.isSuccess) {
                // Reload modules to update the UI
                loadModules()
            } else {
                _error.value = result.exceptionOrNull()?.message ?: "Failed to toggle module"
            }
        }
    }

    fun enableModule(moduleName: String) {
        viewModelScope.launch {
            val result = repository.toggleModule(moduleName, "enable")
            if (result.isSuccess) {
                loadModules()
            } else {
                _error.value = result.exceptionOrNull()?.message ?: "Failed to enable module"
            }
        }
    }

    fun disableModule(moduleName: String) {
        viewModelScope.launch {
            val result = repository.toggleModule(moduleName, "disable")
            if (result.isSuccess) {
                loadModules()
            } else {
                _error.value = result.exceptionOrNull()?.message ?: "Failed to disable module"
            }
        }
    }

    fun clearError() {
        _error.value = null
    }

    fun startAutoRefresh() {
        if (autoRefreshEnabled) return
        autoRefreshEnabled = true

        viewModelScope.launch {
            while (autoRefreshEnabled) {
                val settings = settingsPreferences.settings.first()
                val interval = settings.refreshInterval

                delay(interval * 1000L)

                if (autoRefreshEnabled) {
                    loadModules()
                }
            }
        }
    }

    fun stopAutoRefresh() {
        autoRefreshEnabled = false
    }
}
